src/Uno.UI/UI/Xaml/Controls/Border/BorderLayerRenderer.wasm.cs

```diff
@@ -1,196 +1,249 @@
 			{
-				element.SetStyle(
-					("border-style", "none"),
-					("border-color", ""),
-					("border-width", ""));
-			}
-			else
-			{
-				var borderWidth = $"{thickness.Top.ToStringInvariant()}px {thickness.Right.ToStringInvariant()}px {thickness.Bottom.ToStringInvariant()}px {thickness.Left.ToStringInvariant()}px";
-				switch (brush)
+				switch (img.Kind)
 				{
-					case SolidColorBrush solidColorBrush:
-						var borderColor = solidColorBrush.ColorWithOpacity;
-						element.SetStyle(
-							("border", ""),
-							("border-style", "solid"),
-							("border-color", borderColor.ToHexString()),
-							("border-width", borderWidth));
-						break;
-					case GradientBrush gradientBrush:
-						var border = gradientBrush.ToCssString(element.RenderSize); // TODO: Reevaluate when size is changing
-						element.SetStyle(
-							("border-style", "solid"),
-							("border-color", ""),
-							("border-image", border),
-							("border-width", borderWidth),
-							("border-image-slice", "1"));
-						break;
-					case AcrylicBrush acrylicBrush:
-						var acrylicFallbackColor = acrylicBrush.FallbackColorWithOpacity;
-						element.SetStyle(
-							("border", ""),
-							("border-style", "solid"),
-							("border-color", acrylicFallbackColor.ToHexString()),
-							("border-width", borderWidth));
+					case ImageDataKind.Empty:
+					case ImageDataKind.Error:
+						_owner.ResetStyle("background-color", "background-image", "background-size");
 						break;
+
+					case ImageDataKind.DataUri:
+					case ImageDataKind.Url:
 					default:
-						element.ResetStyle("border-style", "border-color", "border-image", "border-width");
+						_owner.SetStyle(
+							("background-color", ""),
+							("background-origin", "content-box"),
+							("background-position", imgBrush.ToCssPosition()),
+							("background-size", imgBrush.ToCssBackgroundSize()),
+							("background-image", "url(" + img.Value + ")"),
+							("background-repeat", "no-repeat")
+						);
 						break;
 				}
-			}
+			});
 		}
+		else if (brush is AcrylicBrush acrylicBrush)
+		{
+			return acrylicBrush.Subscribe(_owner);
+		}
+		else
+		{
+			return Brush.AssignAndObserveBrush(brush, _ => SetBackground(brush));
+		}
+	}
 
-		public static IDisposable SetAndObserveBackgroundBrush(FrameworkElement element, Brush brush)
+	private void SetCornerRadius(CornerRadius cornerRadius)
+	{
+		if (cornerRadius == CornerRadius.None)
 		{
-			SetBackgroundBrush(element, brush);
+			_owner.ResetStyle("border-radius", "overflow");
+		}
+		else
+		{
+			var borderRadiusCssString = $"{cornerRadius.TopLeft.ToStringInvariant()}px {cornerRadius.TopRight.ToStringInvariant()}px {cornerRadius.BottomRight.ToStringInvariant()}px {cornerRadius.BottomLeft.ToStringInvariant()}px";
+			_owner.SetStyle(
+				("border-radius", borderRadiusCssString),
+				("overflow", "hidden")); // overflow: hidden is required here because the clipping can't do its job when it's non-rectangular.
```

Comment:
I'm wondering if we should create specialized methods in TypeScript for calls from this method instead of relying on the formatter + parsing. Since it's something done pretty often in applications...

Another way would be to move part of the logic to TypeScript instead of managed code.

Reply:
Added new `WindowManagerInterop` method for setting border radius specifically